cachi2/core/package_managers/metayarn.py

```diff
@@ -0,0 +1,122 @@
+
+_yarn_versions = {
+    "yarn_classic": contains_yarn_classic,
+    "yarnberry": contains_yarnberry,
+}
+_yarn_processors = {
+    "yarn_classic": yarn_classic.fetch_yarn_source,
+    "yarnberry": yarn.fetch_yarn_source,
+}
+
+
+def _yarn_selector(
+    package: PackageInput,
+    source_dir: RootedPath,
+) -> tuple[str, Optional[Exception]]:
+    try:
+        for yarn_version, version_matches_for in _yarn_versions.items():
+            if version_matches_for(package, source_dir):
+                return yarn_version, None
+        else:
+            return "uncategorized", None
+    except Exception as e:
+        return "exception", e
+
+
+def _separate_packages_by_yarn_version(
+    packages: Iterable[PackageInput],
+    source_dir: RootedPath,
+) -> dict[str, Union[list[PackageInput], tuple[PackageInput, Exception]]]:
+    """Sorts packages to bins depending on which Yarn version was used.
+
+    The output dictionary contains "uncategorized" entry to capture anything
+    that could not be categorized (likely yet-unsupported versions of Yarn).
+
+    The output dictionary also contains category "exceptions" to accumulate
+    exceptions that occured during pre-processing of packages.
+    """
+    output = defaultdict(list)
+    for p in packages:
+        category, exception = _yarn_selector(p, source_dir)
+        if exception is None:
+            output[category].append(p)
+        else:  # This is an exceptional result.
+            # Categories will never clash, but mypy does not know that and
+            # does not want to learn.
+            output[category].append((p, exception))  # type: ignore
+    return output  # type: ignore
+
+
+def dispatch_to_correct_fetcher(request: Request) -> RequestOutput:
+    """Dispatch a request to correct yarn backend.
+
+    In order to save a user from the need to distinguish between different
+    flavors of Yarn this function attempts to separate Yarn packages and process each
+    with an appropriate manager.
+    """
+    sorted_packages = _separate_packages_by_yarn_version(request.packages, request.source_dir)
+    if uncat := sorted_packages.pop("uncategorized", False):
+        log.warning(f"Failed to categorize the following packages: {uncat}")
+    if exceptions := sorted_packages.pop("exceptions", False):
+        log.warning(f"Following packages caused categorizer to fail: {exceptions}")
+
+    fetched_packages = []
+    for pm, packages in sorted_packages.items():
+        new_request = request.model_copy(update={"packages": packages})
+        fetched_packages.append(_yarn_processors[pm](new_request))
+    # Judging from the resolver code it is safe to merge multiple packages.
+    # Moreover, it does not look like there is any mechanism in place now to
+    # prevent users from requesting both PMs simultaneously.
+    # The code below preserves this behavior.
+    return _merge_outputs(fetched_packages)
```

Comment:
How do plan on plugging this to the existing `_resolve_packages` helper in the `resolver` module?

Reply:
`_merge_outputs` takes `Iterable[RequestOutput]` and returns `RequestOutput`. Each `fetch_foo_deps` returns a `RequestOutput` so merging any two or more of those in any order should be fine (if I understand your question correctly).

Reply:
Stupid me didn't look at the return type, just the input type. In that case it should be simple.